src/ModuleHomomorphism.jl

```diff
@@ -124,38 +111,37 @@ an exception is raised.
 """
 function preimage(f::Map(FPModuleHomomorphism), v::FPModuleElem{T}) where
                                                                T <: RingElement
+   return preimage(f, [v])[1]
+end
+
+function preimage(f::Map(FPModuleHomomorphism), v::Vector{<:FPModuleElem{T}}) where
+                                                               T <: RingElement
    D = domain(f)
    C = codomain(f)
    R = base_ring(C)
-   parent(v) !== C && error("Incompatible element")
+   if length(v) == 0
+      return elem_type(domain(f))[]
+   end
+   parent(v[1]) !== C && error("Incompatible element")
    M = matrix(f)
    trels = rels(C)
    # Put rows of M and target relations into a matrix
    q = length(trels)
    m = nrows(M)
    n = ncols(M)
-   ncols(Generic._matrix(v)) != n && error("Incompatible element")
    if m == 0 || n == 0
-       return D(zero_matrix(R, 1, m))
+      return elem_type(D)[D(zero_matrix(R, 1, m)) for x in v]
    else
       # Put matrix M and target relations in a matrix
       matr = zero_matrix(R, m + q, n)
-      for i = 1:m
-         for j = 1:n
-            matr[i, j] = M[i, j]
-         end
-      end
+      matr[1:m, 1:n] = M
       for i = 1:q
-         for j = 1:n
-            matr[m + i, j] = trels[i][1, j]
-         end
+        matr[m + i, :] = trels[i]
       end
       # Find left inverse of mat
-      x = solve(matr, Generic._matrix(v))
-      if q != 0
-         x = matrix(R, 1, m, T[x[1, i] for i in 1:m])
-      end
-      return D(x)
+      inmat = reduce(vcat, Generic._matrix.(v))
+      x = solve(matr, inmat)
+      return elem_type(D)[D(x[i:i, 1:m]) for i in 1:length(v)]
    end
 end
 
```

Comment:
For future use (I am happy withthis change as is): should we put a SolveCtx in the map? In some ways then the detour via preimage of a Vector is then not neccessary...